### DISCUSSION TITLE
Unable to use more conformations

### DISCUSSION THREAD

**Discussion by ZhongyuWAN** [SSCHAcode/python-sscha#339] [Q&A] [answered]
hello, SSCHA Developers.

I have recently been utilizing SSCHA to conduct an NVT simulation on a structure of interest, the pressure is ambient.

Under the harmonic approximation, the structure exhibits dynamical stability as there are no imaginary frequencies. However, upon introducing anharmonic corrections, the simulation terminates prematurely during the initial phase. A file named "error_struct.scf" is generated in the directory.

Here is the error message:

> Force computed shape: 800
Computing configuration 1 out of 200 (nat = 56)
TOTAL INPUT:
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_0'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
TOTAL INPUT:
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_1'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
TOTAL INPUT:
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_3'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
TOTAL INPUT:
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_2'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
Rerun the job 0
TOTAL INPUT:
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_0'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
TOTAL INPUT:
TOTAL INPUT:
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_3'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_2'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
TOTAL INPUT:
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_1'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
Rerun the job 0
TOTAL INPUT:
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_0'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
TOTAL INPUT:
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_2'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
TOTAL INPUT:
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_3'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
TOTAL INPUT:
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_1'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
Rerun the job 0
TOTAL INPUT:
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_0'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
TOTAL INPUT:
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_2'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
TOTAL INPUT:
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_3'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
TOTAL INPUT:
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_1'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
Rerun the job 0
TOTAL INPUT:
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_0'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
TOTAL INPUT:
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_2'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
TOTAL INPUT:
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_3'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
TOTAL INPUT:
{'control': {'disk_io': 'None', 'pseudo_dir': '.', 'tprnfor': True, 'tstress': True, 'prefix': 'esp_1'}, 'system': {'ecutwfc': 70, 'ecutrho': 280, 'occupations': 'smearing', 'smearing': 'gauss', 'degauss': 0.04, 'vdw_corr': 'grimme-d2', 'nat': 56, 'ntyp': 2, 'ibrav': 0}, 'electrons': {'conv_thr': 1e-10, 'diagonalization': 'david', 'mixing_mode': 'plain', 'mixing_beta': 0.7}}
Rerun the job 0
Error in the ASE calculator for more than 5 times
     while computing 'error_struct.scf'Traceback (most recent call last):
  File "/gpfs1/scratch/zywan5/anharm/Rb/0.04-222/anharmonic/anha-NVT/NVT-2/sscha-anhamonic.py", line 101, in <module>
    relax.relax(get_stress = True)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/sscha/Relax.py", line 353, in relax
    self.minim.ensemble.compute_ensemble(self.calc, get_stress,
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/sscha/Ensemble.py", line 3156, in compute_ensemble
Error in the ASE calculator for more than 5 times
     while computing 'error_struct.scf'Traceback (most recent call last):
  File "/gpfs1/scratch/zywan5/anharm/Rb/0.04-222/anharmonic/anha-NVT/NVT-2/sscha-anhamonic.py", line 101, in <module>
    relax.relax(get_stress = True)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/sscha/Relax.py", line 353, in relax
    self.minim.ensemble.compute_ensemble(self.calc, get_stress,
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/sscha/Ensemble.py", line 3156, in compute_ensemble
Error in the ASE calculator for more than 5 times
     while computing 'error_struct.scf'Traceback (most recent call last):
  File "/gpfs1/scratch/zywan5/anharm/Rb/0.04-222/anharmonic/anha-NVT/NVT-2/sscha-anhamonic.py", line 101, in <module>
    relax.relax(get_stress = True)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/sscha/Relax.py", line 353, in relax
Error in the ASE calculator for more than 5 times
     while computing 'error_struct.scf'    self.minim.ensemble.compute_ensemble(self.calc, get_stress,
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/sscha/Ensemble.py", line 3156, in compute_ensemble
Traceback (most recent call last):
  File "/gpfs1/scratch/zywan5/anharm/Rb/0.04-222/anharmonic/anha-NVT/NVT-2/sscha-anhamonic.py", line 101, in <module>
    relax.relax(get_stress = True)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/sscha/Relax.py", line 353, in relax
    self.minim.ensemble.compute_ensemble(self.calc, get_stress,
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/sscha/Ensemble.py", line 3156, in compute_ensemble
    computing_ensemble.get_energy_forces(calculator, compute_stress, stress_numerical, verbose = verbose)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/sscha/Ensemble.py", line 3416, in get_energy_forces
    computing_ensemble.get_energy_forces(calculator, compute_stress, stress_numerical, verbose = verbose)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/sscha/Ensemble.py", line 3416, in get_energy_forces
    computing_ensemble.get_energy_forces(calculator, compute_stress, stress_numerical, verbose = verbose)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/sscha/Ensemble.py", line 3416, in get_energy_forces
    computing_ensemble.get_energy_forces(calculator, compute_stress, stress_numerical, verbose = verbose)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/sscha/Ensemble.py", line 3416, in get_energy_forces
    results = CC.calculators.get_results(ase_calculator, struct, get_stress = compute_stress)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/cellconstructor/calculators.py", line 84, in get_results
    results = CC.calculators.get_results(ase_calculator, struct, get_stress = compute_stress)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/cellconstructor/calculators.py", line 84, in get_results
    results = CC.calculators.get_results(ase_calculator, struct, get_stress = compute_stress)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/cellconstructor/calculators.py", line 84, in get_results
    results = CC.calculators.get_results(ase_calculator, struct, get_stress = compute_stress)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/cellconstructor/calculators.py", line 84, in get_results
    calculator.calculate(structure)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/cellconstructor/calculators.py", line 113, in calculate
    calculator.calculate(structure)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/cellconstructor/calculators.py", line 113, in calculate
    calculator.calculate(structure)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/cellconstructor/calculators.py", line 113, in calculate
    calculator.calculate(structure)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/cellconstructor/calculators.py", line 113, in calculate
    self.execute()
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/cellconstructor/calculators.py", line 135, in execute
    self.execute()
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/cellconstructor/calculators.py", line 135, in execute
    self.execute()
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/cellconstructor/calculators.py", line 135, in execute
    self.execute()
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/site-packages/cellconstructor/calculators.py", line 135, in execute
    proc = subprocess.Popen(cmd, shell = True, env = new_env, cwd = self.directory, stdout = foutput)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/subprocess.py", line 951, in __init__
    proc = subprocess.Popen(cmd, shell = True, env = new_env, cwd = self.directory, stdout = foutput)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/subprocess.py", line 951, in __init__
    proc = subprocess.Popen(cmd, shell = True, env = new_env, cwd = self.directory, stdout = foutput)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/subprocess.py", line 951, in __init__
    proc = subprocess.Popen(cmd, shell = True, env = new_env, cwd = self.directory, stdout = foutput)
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/subprocess.py", line 951, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/subprocess.py", line 1821, in _execute_child
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/subprocess.py", line 1821, in _execute_child
    self._execute_child(args, executable, preexec_fn, close_fds,
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/subprocess.py", line 1821, in _execute_child
  File "/gpfs1/home/zywan5/miniconda3/lib/python3.9/subprocess.py", line 1821, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
    raise child_exception_type(errno_num, err_msg, err_filename)
OSError: [Errno 14] Bad address: '/bin/sh'
    raise child_exception_type(errno_num, err_msg, err_filename)
    raise child_exception_type(errno_num, err_msg, err_filename)
OSError: [Errno 14] Bad address: '/bin/sh'
OSError: [Errno 14] Bad address: '/bin/sh'
OSError: [Errno 14] Bad address: '/bin/sh'

After adjusting the parameter N_CONFIGS from 800 to 80, the SSCHA calculations were able to run smoothly and eventually outputted the dynamical matrix.

Therefore, I am wondering why using a larger N_CONFIGS value leads to the program not functioning properly. I would appreciate any suggestions on how I can improve this issue.

Zhongyu

**Comment by mesonepigreco** [comment]
Hi, the quantum espresso calculation was failing. After 5 failed attempts, the code stopped.
Maybe there is a memory problem? The file error_struct.scf is the structure of the espresso calculation that was failing.

The code should also have created the input/output of the failing espresso calculation; you can check it there to better understand the problem. However,if you run again, it has been overwritten.
Try again with the high number of configurations,and check the specific error espresso is giving

**Reply by ZhongyuWAN** [reply]
Hello, mesonepigreco

Thank you very much for your response. In the SCF calculation for a single conformation, the QE output file indicates: "Estimated max dynamical RAM per process > 115.67 MB, Estimated total dynamical RAM > 1.81 GB."

Given the capabilities of my current machine, it can handle 128 cores and 450GB of RAM. However, the introduction of more conformations seems to exceed the computational limits of the machine. Nevertheless, using more conformations appears necessary as a small number of conformations does not seem to allow the FC gradient to approach zero.

Are there any alternative solutions besides upgrading the computational equipment? For instance, could we perform the SCF calculation using only one k-point or utilize a smaller unit cell?

Another point of confusion for me is that in some works utilizing SSCHA, such as in Journal of Applied Physics 130.17 (2021) and Physical Review B 103.13 (2021): 134305, it seems they did not test for size effects by using larger unit cells. I understand that larger unit cells can lead to significantly longer computation times. 

However, is the information provided solely in Fig. 2 of Tutorial-1 sufficient to judge the convergence of the calculations? Is it necessary to check for size effects by varying the size of the unit cell? Especially when considering the inability to perform calculations with larger supercell sizes.

Zhongyu

**Reply by mesonepigreco** [reply, marked as answer]
Hi, 
So here there are two distinct problems:
1) The number of configurations. This should not impact that much on the resources, as it scales linearly (just how many SCF calculations are needed). 
2) The dimension of the supercell. This scales much worse, as both the single calculation scales at least as N^3 as the number of calculations needed. 

Regarding 1, I believe the problem is more related to the fact that you are probably submitting all the calculations in one single node. If you are on a cluster, the best approach should be using the Cluster module, as shown in the [second tutorial](https://sscha.eu/Tutorials/tutorial_02_advanced_submission/) on the website. This would avoid memory conflicts between different calculations and allow you to exploit the cluster architecture much better. If it is not a cluster, try again with the big number of random configurations and look at the failing QE calculation why it is crashing. It should not.

Regarding 2: Yes, you have to converge with the supercell size. Nowadays, this is becoming easier and easier thanks to the availability of machine-learning force fields ([tutorial 6](https://sscha.eu/Tutorials/tutorial_06_the_SSCHA_with_MLP/)). So the solution usually is to study the SCHA on a small cell, then use the data generated during the sscha simulation to train an ML potential  ([tutorial 6](https://sscha.eu/Tutorials/tutorial_06_the_SSCHA_with_MLP/)), then run the SSCHA on a bigger cell until converges.

**Comment by yw-fang** [comment]
It seems that it's just because some of the supercell calculations failed in Quantum Espresso. 
This error didn't rely on the number of supercells used in the population. The SSCHA calculation didn't show error in the case with 80 supercells were because all calculations were finished successfully in QE. You can check the status of all the QE calculations in the calculations with 800 supercells to confirm if there were jobs failed.

**Reply by ZhongyuWAN** [reply]
hello, Yue wen

Thank you for your advice. As mesonepigreco speculated, it seems that 800 configurations were occupying the memory simultaneously, but I'm curious about the reason behind this occurrence. Although I have resolved the issue using the cluster module, I still want to provide you with a reference.

Theoretically, running the Python script from Tutorial-1 in parallel on the machine should not have led to such an issue. The command to invoke QE is command = 'mpirun -np 32 pw.x -i PREFIX.pwi > PREFIX.pwo' within the SSCHA script. And in the script for submitting jobs, I used mpirun -np 4 python sscha-anhamonic.py. This indicates that 128 cores are being utilized simultaneously, with 4 pw.x calculations running, each calling 32 cores.

Zhongyu

**Reply by mesonepigreco** [reply]
This could not be optimal, as clusters can map processes into real cores to avoid multiple processes running on the same cores. This binding is usually done by running ``srun``, so you should link the compilation of mpi4py (the MPI wrapper in Python employed by the SSCHA) with the mpi version provided by the cluster. However, the procedure could be extremely complex, so it is often much better to use the cluster scheduler.
